fauna/models.py

```diff
@@ -16,43 +17,130 @@ def __hash__(self):
+        super().__init__(coll)
+        self._id = ref_id
+
+    def __hash__(self):
+        hash((type(self), self._collection, self._id))
 
     def __str__(self):
-        return f"{self.collection}:{self.ref_id}"
+        return f"{self._collection}:{self._id}"
 
     def __eq__(self, other):
-        return isinstance(other, DocumentReference) and str(self) == str(other)
+        return isinstance(other, type(self)) and str(self) == str(other)
 
-    def __hash__(self):
-        hash((self.collection, self.ref_id))
+    @property
+    def id(self) -> int:
+        return self._id
 
     @staticmethod
     def from_string(ref: str):
         rs = ref.split(":")
         if len(rs) != 2:
-            raise ValueError(
-                "Expects string of format <CollectionName>:<RefID>")
-        return DocumentReference(rs[0], rs[1])
+            raise ValueError("Expects string of format <CollectionName>:<ID>")
+        return DocumentReference(rs[0], int(rs[1]))
 
 
-class DocumentDict(dict):
+class NamedDocumentReference(BaseReference):
+    _name: str
 
-    def ref(self) -> DocumentReference:
-        for k in ["coll", "id"]:
-            if k not in self:
-                raise ValueError(
-                    f"Document does not contain the required '{k}' key to return a reference"
-                )
+    def __init__(self, coll: Union[str, Module], name: str):
+        super().__init__(coll)
+        self._name = name
+
+    def __hash__(self):
+        hash((type(self), self._collection, self._name))
+
+    def __str__(self):
+        return f"{self._collection}:{self._name}"
+
+    def __eq__(self, other):
+        return isinstance(other, type(self)) and str(self) == str(other)
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+
+T = TypeVar('T')
+
+
+class BaseDocument(Mapping[str, T]):
+
+    def __init__(self, *args, **kwargs):
+        self._store = dict(*args, **kwargs)
 
+    def __getitem__(self, __k: str) -> T:
+        return self._store[__k]
+
+    def __len__(self) -> int:
+        return len(self._store)
+
+    def __iter__(self) -> Iterator[T]:
+        return iter(self._store)
+
+    def __eq__(self, other):
+        if not isinstance(other, type(self)):
+            return False
+
+        if len(self) != len(other):
+            return False
+
+        for k, v in self.items():
+            if k not in other:
+                return False
+            if self[k] != other[k]:
+                return False
+
+        return True
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+
+class Document(BaseDocument):
+
+    def __init__(self, data: dict):
+        if "coll" not in data or "id" not in data:
+            raise ValueError("Data must contain the 'coll' and 'id' keys")
+
+        super().__init__(data)
+
+    @property
+    def ref(self) -> DocumentReference:
         return DocumentReference(self["coll"], self["id"])
```

Comment:
`self["id"]` will be a string, and needs to be converted. This will only fail for an invalid document.
```suggestion
        return DocumentReference(self["coll"], int(self["id"]))
```

Reply:
Since we're back to storing a string, I'll leave this out